crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+    /// `scan_limit` if `is_from_front` is false.
+    pub(crate) fn scan_lo(&self) -> u64 {
+        let adjusted_lo = self.tx_lo();
+
+        if self.is_from_front {
+            adjusted_lo
+        } else {
+            // If not from the front, then the scan_limit must only be applied to the lower bound
+            if let Some(scan_limit) = self.scan_limit {
+                adjusted_lo.max(self.tx_hi().saturating_sub(scan_limit))
+            } else {
+                adjusted_lo
+            }
+        }
+    }
+
+    /// The upper bound `tx_sequence_number` of the range to scan within. This defaults to the max
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the upper bound is
+    /// adjusted to the smaller of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is true.
+    pub(crate) fn scan_hi(&self) -> u64 {
+        let adjusted_hi = self.tx_hi();
+
+        if self.is_from_front {
+            if let Some(scan_limit) = self.scan_limit {
+                adjusted_hi.min(self.tx_lo().saturating_add(scan_limit))
+            } else {
+                adjusted_hi
+            }
+        } else {
+            adjusted_hi
+        }
+    }
+
+    /// If the query result does not have a previous page, check whether the scan limit is within
+    /// the initial tx_sequence_number range.
+    pub(crate) fn scan_has_prev_page(&self) -> bool {
+        if self.after.unwrap_or(0) >= self.hi {
+            return false;
+        }
+
+        self.scan_lo() > self.lo
+    }
+
+    /// If the query result does not have a next page, check whether the scan limit is within the
+    /// initial tx_sequence_number range.
+    pub(crate) fn scan_has_next_page(&self) -> bool {
+        if self.before.unwrap_or(self.hi) <= self.lo {
+            return false;
+        }
+
+        self.scan_hi() < self.hi
+    }
+
+    pub(crate) fn scan_prev_page_and_lo(&self) -> (bool, u64) {
+        let has_prev = self.scan_has_prev_page();
+
+        (has_prev, self.scan_lo())
+    }
+
+    pub(crate) fn scan_next_page_and_hi(&self) -> (bool, u64) {
+        let has_next = self.scan_has_next_page();
+
+        (has_next, self.scan_hi())
+    }
+}
+
+impl TransactionBlockFilter {
+    /// Most filter conditions require a scan limit if used in tandem with other filters. The
+    /// exception to this is sender and checkpoint, since sender is denormalized on all tables, and
+    /// the corresponding tx range can be determined for a checkpoint.
+    pub(crate) fn requires_scan_limit(&self) -> bool {
+        [
+            self.function.is_some(),
+            self.kind.is_some(),
+            self.recv_address.is_some(),
+            self.input_object.is_some(),
+            self.changed_object.is_some(),
+            self.transaction_ids.is_some(),
+        ]
+        .into_iter()
+        .filter(|is_set| *is_set)
+        .count()
+            > 1
```

Comment:
```suggestion
        [
            &self.function,
            &self.kind,
            &self.recv_address,
            &self.input_object,
            &self.changed_object,
            &self.transaction_ids,
        ]
        .into_iter()
        .any(|f| f.is_some())
```

Reply:
`.any(|f| f.is_some())` will be true if one of the filters is set but iiuc we actually want this function to return true if at least two filters are set.

Reply:
yeah agree with Emma here